File: app/controllers/continuity.rb
class Continuity

  def cos_sim(text)
    sentence_arr = text.split('.')

    sentence_arr.map! { |sentence| clean_text(sentence) }
    sentence_arr.delete([])

    return 0 if sentence_arr.length < 2

    result_arr = Array.new(sentence_arr.length - 1)
    sentence_arr.each_index do |i|
      unless i == sentence_arr.length - 1
        sentence1 = sentence_arr[i]
        sentence2 = sentence_arr[i+1]

        type_array = []

        sentence1.each do |word|
          unless type_array.include?(word)
            type_array << word
          end
        end
        sentence2.each do |word|
          unless type_array.include?(word)
            type_array << word
          end
        end

        vector1 = Array.new(type_array.size, 0)
        vector2 = Array.new(type_array.size, 0)

        sentence1.each do |word|
          vector1[type_array.find_index(word)] += 1
        end
        sentence2.each do |word|
          vector2[type_array.find_index(word)] += 1
        end

        normalize!(vector1)
        normalize!(vector2)

        result_arr[i] = dot_product(vector1, vector2)
      end
    end

    result_arr.reduce(:+) / result_arr.length
  end

  def clean_text(text)
    new_text = text.gsub(/<(.|\n)*?>/, ' ')
    new_text = new_text.gsub(/&nbsp;/, ' ')
    new_text = new_text.gsub(/\\n/, ' ')
    new_text = new_text.gsub(/[^a-z0-9 ]/i, '')
    new_text = new_text.downcase
    new_text.split
  end

  def normalize!(vector)
    length = 0
    vector.each { |element| length += element * element }
    length = Math.sqrt(length)
    vector.map! { |element| element / length }
    vector
  end

  def dot_product(vector1, vector2)
    result = 0
    vector1.each_index { |i| result += vector1[i] * vector2[i] }
    result
  end
end

File: app/controllers/continuity.js
a = function (text) {

  var cos_sim = function (text) {
    var sentence_arr = text.match(/[^.]+/gi);

    sentence_arr = sentence_arr.map(clean_text);

    for (var i = sentence_arr.length-1; i >= 0; i--) {
      if (sentence_arr[i] === []) {
          sentence_arr.splice(i, 1);
      }
    }

    var result_arr = new Array(sentence_arr.length - 1)

    for (var i = 0; i < sentence_arr.length; i++) {
      if (i !== sentence_arr.length - 1) {
        var sentence1 = sentence_arr[i];
        var sentence2 = sentence_arr[i+1];

        var type_array = [];

        for (var j = 0; j < sentence1.length; j++) {
          if (type_array.indexOf(sentence1[j]) === -1) {
            type_array.push(sentence1[j]);
          }
        }
        for (var j = 0; j < sentence2.length; j++) {
          if (type_array.indexOf(sentence2[j]) === -1) {
            type_array.push(sentence2[j]);
          }
        }

        var vector1 = new Array(type_array.length);
        var vector2 = new Array(type_array.length);

        for (var j = 0; j < vector1.length; j++) {
          vector1[j] = 0;
          vector2[j] = 0;
        }

        for (var j = 0; j < sentence1.length; j++) {
          vector1[type_array.indexOf(sentence1[j])]++;
        }
        for (var j = 0; j < sentence2.length; j++) {
          vector2[type_array.indexOf(sentence2[j])]++;
        }

        normalize(vector1);
        normalize(vector2);

        result_arr[i] = dot_product(vector1, vector2);
      }
    }

    var total = 0;
    for (var i = 0; i < result_arr.length; i++) {
      total += result_arr[i];
    }
    return total / result_arr.length;
  }

  var normalize = function (vector) {
    var length = 0;
    for (var i = 0; i < vector.length; i++) {
      length += vector[i] * vector[i];
    }
    length = Math.sqrt(length);
    for (var i = 0; i < vector.length; i++) {
      vector[i] = vector[i] / length;
    }
    return vector;
  }

  var dot_product = function (vector1, vector2) {
    var result = 0;
    for (var i = 0; i < vector1.length; i++) {
      result += (vector1[i] * vector2[i]);
    }
    return result;
  }

  var clean_text = function (text) {
    var new_text = text.replace(/[^\w\s\d]/gi, '');
    return new_text.toLowerCase().match(/[a-z0-9]+/gi);
  }

  return cos_sim(text);
}